### ISSUE TITLE
upgrade to ksonnet 0.12.0, jsonnet v0.11.2 in the Docker image for our test workers

### ISSUE THREAD

**Issue by kkasravi** [kubeflow/kubeflow#1540]
issue #1534 requires a version of workflows/k8s.libsonnet that includes apiextensions
This would be in {ksonnet 0.12.0, jsonnet v0.11.2}

Are we able to upgrade to ksonnet 0.12.0? 
I wasn't able figure out where to upgrade ksonnet other that what was already done 
for the web backend #1508

**Comment by jlewi** [comment]
The ksonnet version is installed in our test image.
https://github.com/kubeflow/testing/blob/master/images/Dockerfile#L92

So we'd need to update that and then build and push a new version of that image.

I'm unclear though why #1534 requires a ksonnet update.

**Comment by jlewi** [comment]
I think this is one PR where we were discussing this
https://github.com/kubeflow/kubeflow/pull/1437#issuecomment-418523083

**Comment by kkasravi** [comment]
You had said that for some of the libraries which aren't normally user-facing, the use of k.libsonnet would be ok. Others that have exposure to non-ksonnet experts we shouldn't go overboard on k.libsonnet, k8s.libsonnet.

**Comment by kkasravi** [comment]
I can update the Dockerfile and submit it as a PR to this issue

**Comment by kkasravi** [comment]
see https://github.com/kubeflow/testing/pull/201

**Comment by jlewi** [comment]
> You had said that for some of the libraries which aren't normally user-facing, the use of k.libsonnet would be ok. Others that have exposure to non-ksonnet experts we shouldn't go overboard on k.libsonnet, k8s.libsonnet.

You are correct. I still am uncertain about what the best practice/style for using k.libsonnet is.

Your PR #1535 was helpful to spark some initial thoughts. See comment:
https://github.com/kubeflow/kubeflow/pull/1535#issuecomment-422177056

**Comment by jlewi** [comment]
kubeflow/testing#201 is merged I'll build a new Docker image with the changes.

**Comment by jlewi** [comment]
I built new docker images
gcr.io/kubeflow-ci/test-worker:v20180917-b094114-e3b0c4 
gcr.io/kubeflow-releasing/worker:v20180917-b094114-e3b0c4 

These are now tagged latest as well.

**Comment by jlewi** [comment]
Looks like this caused a bunch of presubmit failures; error message is

```
INFO|2018-09-18T00:51:05|/src/kubeflow/testing/py/kubeflow/testing/util.py|62| time="2018-09-18T00:51:05Z" level=warning msg="Versioned packages stored in unversioned paths - please run `ks upgrade` to correct."
```

**Comment by jlewi** [comment]
Looks like the previous images were
gcr.io/kubeflow-ci/test-worker:v20180822-da7cefc-e3b0c4
gcr.io/kubeflow-releasing/worker:v20180822-da7cefc-e3b0c4

Adding the latest tag back to those images.

see: kubeflow/testing#196

**Comment by jlewi** [comment]
It looks like if we want to upgrade to ksonnet 0.12 we will need to upgrade to the ksonnet app used for our tests.

I think the way to handle this is we can upgrade individual tests to change the test-worker image to use the image
gcr.io/kubeflow-ci/test-worker:v20180917-b094114-e3b0c4

rather than the "latest" tag. This way the presubmit will verify the tests pass with the new image.

Ideally, the test worker image should be a ksonnet parameter that we set it in the prow_config.yaml
https://github.com/kubeflow/kubeflow/blob/master/prow_config.yaml.

The reason for setting it in prow_config.yaml and not the jsonnet file is that makes it easier to see its consistently set across all E2E tests. It looks like a lot of our E2E tests though don't expose it as a parameter.

An example is the tf-operator E2E tests
https://github.com/kubeflow/tf-operator/blob/master/prow_config.yaml#L17

**Comment by kunmingg** [comment]
I feel that ideally our test workflows shouldn't have dependency on ks versions.
How about turn test workflows into prototypes and e2e test will generate from them on the fly?

**Comment by jlewi** [comment]
@kunmingg isn't dynamically creating ksonnet apps going to be a lot more difficult?

run_e2e_workflow.py can figure out which ksonnet version to use just by looking at app.yaml in the app dir specified in prow_config.yaml.We can install multiple versions of ksonnet in the same Docker image e.g
ks-11
ks-12

This seems like a fairly tractable way to deal with ksonnet's app versioning.

**Comment by kkasravi** [comment]
@jlewi @kunmingg @ankushagarwal seems like kfctl.sh also needs this logic 

```
function check_install() {
  if ! which "${1}" &>/dev/null; then
    echo "You don't have ${1} installed. Please install ${1}."
    exit 1
  fi
}
# TODO(ankushagarwal): verify ks version is higher than 0.11.0
check_install ks
```
and should ask the user to install the latest ks version.
